src/Compilers/Core/CommandLine/CompilerServerLogger.cs

```diff
@@ -8,11 +8,86 @@
 using System.Diagnostics;
 using System.IO;
 using System.Linq;
+using System.Runtime.InteropServices.WindowsRuntime;
 using System.Text;
 using System.Threading;
 
 namespace Microsoft.CodeAnalysis.CommandLine
 {
+    /// <summary>
+    /// Used to log information from within the compiler server
+    /// </summary>
+    /// <remarks>
+    /// Implementations of this interface must assume they are used on multilpe threads without any form
+    /// of synchronization.
+    /// </remarks>
+    internal interface ICompilerServerLogger
+    {
+        bool IsLogging { get; }
+        void Log(string message);
+    }
+
+    internal static class CompilerServerLoggerExtensions
+    {
+        internal static void Log(this ICompilerServerLogger logger, string format, params object?[] arguments)
+        {
+            if (logger.IsLogging)
+            {
+                logger.Log(string.Format(format, arguments));
+            }
+        }
+
+        internal static void LogError(this ICompilerServerLogger logger, string message)
```

Comment:
Do we want to not have this overload, to encourage not using it and always passed arguments on the side?

Reply:
Went back and forth on this. At the end of the day though the allocations the interpolated strings never really show up in a performance trace hence I error'd on the side of usability here.